### PR TITLE
kernel-install changes to bootloader module

### PR DESCRIPTION
**Pull request by dalto8**
We made some changes for EndeavourOS to support systemd-boot/kernel-install and we thought we should make them available here.

These changes have the following effects/benefits:
* Produce entries and use install locations compatible with kernel-install as provided by systemd
* Support for multiple kernels to be installed/available
* Support for additional initrd images such ucode files required by Arch

There are some downsides as well
* Some of the customization options were removed to produce something compatible with kernel-install
* There are two pieces of this that are Arch specific that I didn't see an obvious way to generalize
  * The generation of the initramfs requires the kernel version to be passed to it and calls mkinitcpio directly
  * The way that we are getting the kernel version is also specific to the file layout in Arch

I am not really sure this is better or worse than what was there before.  It is just a different approach.

### CODE REVIEW

**Comment by killajoe** [comment]
Tested already on EndeavourOS:
https://github.com/endeavouros-team/ISO/releases/download/0-EndeavourOS-development-ISO-releases/endeavouros-devel-SDB-2021.02.26-x86_64.iso
Includes this changes.

File: src/modules/bootloader/bootloader.schema.yaml
Comment (adriaandegroot) [inline src/modules/bootloader/bootloader.schema.yaml:11]: Because this changes the schema -- so distro's need to go over their configs extra-carefully -- I'm tagging this as a 3.3 pull-request. For a distro like EndeavourOS that knows what it is doing and **chooses** to pull this in, updating the config is natural, but I don't want to surprise other consumers with it.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: There's not much point in expanding bad documentation with more filler. If you're going to add `:param` lines, please be better than previous authors and document what the parameters really do (presumably both are strings, used for .. something)

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: Since this is a 3.3 PR **anyway**, you might want to look at Python's f-strings as well. That makes string-formatting and concatenation just a little easier to read (e.g. `f"{entry_name} {kernel_version}"`) -- easier to read once you're used to it, I suppose.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py:146]: What if the file is not there (yet)? Do we assume the *machineid* module is run before this one, or the installer has some other mechanism to generate the machineid? (You can set some kind of dependencies between modules in `module.desc`)

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: `os.makedirs(machine_dir, exist_ok=True)`

(Not a blocker; could-be-better)

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: .. unless you specifically need the name `machine_dir`, you can skip over that one and go straight to

```
files_dir = os.path.join(install_path, efi_dir, machine_id, kernel_version)
os.makedirs(files_dir, exist_ok=True)
```

to make all the intermediate ones in one go, without worrying about `FileExistsError`. (Assumes *efi_dir* is supposed to be its own directory)

(Not a blocker)

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: This will block the PR though: having to edit Python code for other distro's isn't acceptable.

**Why** is this arch-specific? (It looks like a pretty generic call to mkinitcpio to me) Does the initcpiocfg / initcpio module not do enough for you? Is it essential to do the mkinitcpio **here**, **now**, or could it be done later in a dedicated module?

I'm asking that last question especially because there's configuration options elsewhere that might be able to handle the "arch-specificness" making the rest of this PR acceptable as-is.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: By this point I'd forgotten what `files_dir` was; can we find a better name for that? It is the "EFI path for this bootloader and its files", maybe `efi_dir_for_bootloader_files`? (Not a blocker)

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py:177]: Thank you for taking care with KeyError; many of the Python modules are sensitive to having all their configuration keys, and your code makes using this module in the face of "I don't care about additional Initrd" **much** nicer.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: Maybe we should rename `install_path` to `target_system_root_mountpoint` to make the code more self-documenting.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: That join-and-relpath looks really weird. Er .. maybe not, since `files_dir` is a subdirectory of `install_path` (those are "where is EFI mounted right now in the target system" and "target system root mount" respectively). How about `assert(files_dir.startswith(install_path))` and using `files_dir[len(install_path):]` ? Needs a comment, anyway, that that second path is relative to the target.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py:161]: +1

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: Seems weird to mix `join()` and `+`, plus this has been calculated before -- how about a `target_efi_top_dir` or so? Those `target_`* variables could be placed together so it is clear how the structure works.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: Seeing these long parameter lists makes me thing that kwargs-style might be a more readable approach here. Maybe it's just because the PR review doesn't offer me the convenience of an IDE to help me understand what does what.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py]: The four lines above are used to generate two variables for use in the calls to `create_systemd_boot_conf()`, namely *file* and *rel_root*.  Then there's two variants / flavors, namely "default" and "fallback".

A function returning a list of 3-tuples could factor out some of the Arch-ness: on arch, do this `os.walk()`, elsewhere return a list with just the default flavor (and maybe "fallback").

**Review by adriaandegroot — CHANGES_REQUESTED** [review]
I like it in general, with three points (which you can find in the detailed review comments):
- this is a 3.3-timeframe change because the schema / configuration changes
- blocked by arch-specifics which would need tidying up or making-configurable; some of the arch-specifics might be movable to other modules
- various tidbits and the source could use more overall cleanup (but that's not necessarily on you)

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py]: I suppose it isn't Arch specific.  It is specific to a distro using both `mkinitcpio` and `kernel-install` which is still fairly specific.

It doesn't have to be done here.  It could be done before or after this point.  It needs the kernels to be installed and the machine id to be generated.

As for using the existing initcpio module, the problem would be that it would need to be refactored substantially in order to add the ability to search for all the kernels to know where to write the initrd.  In c++, that would be a decent amount of code.  It would also mean all that logic was duplicated in two places, once in c++ and once in python.

I am not broadly familiar with Calamares outside of this module.  Is there a better way to approach it?

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py]: Yes, that is a much better idea.  I wasn't familiar with os.makedirs.

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py]: That is a great idea.  If we can devise a strategy for resolving the mkinitcpio problem above, there would be nothing left that isn't generic.

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py]: Sorry about that, I tried to stay consistent with what was already there.  I can clean up and document all the params here.

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py:146]: I should add some better error handling there but that wouldn't really resolve the issue since we need the machine-id.

I can add the module dependency if that would be an acceptable solution.

File: src/modules/bootloader/main.py
Comment (dalto8) [reply src/modules/bootloader/main.py]: Yes, that isn't documented very well.  I sort of assumed it would need to be refactored since it is overly specific.

To be clear, the `relpath` is producing a path suitable for use in the chroot by removing `install_path`.  I think it could be made self documenting by making the relpath call a separate statement with a descriptive variable name.

**Comment by adriaandegroot** [comment]
@dalto8 @killajoe Can you tweak / massage this PR so it can go in? I got the impression it was pretty close and would only need some minor twiddles to be more applicable.

In the meantime, it would be nice if you re-target it to the *work-3.3* branch, because that one does not care about Python-version-bumps and things.

**Comment by dalto8** [comment]
> @dalto8 @killajoe Can you tweak / massage this PR so it can go in? I got the impression it was pretty close and would only need some minor twiddles to be more applicable.

Yes.  I will go through it and resolve all the open issues.

> In the meantime, it would be nice if you re-target it to the _work-3.3_ branch, because that one does not care about Python-version-bumps and things.

I can do that.

**Comment by killajoe** [comment]
yes i am in if i can 😌

**Comment by dalto8** [comment]
I think this is ready for review again.

One problem is that I have some commits here from master.  I am not entirely sure how to fix that from a git perspective.

**Comment by kkofler** [comment]
You should rebase your `non-EOS` branch onto `work-3.3` and force-push it (the `non-EOS` branch) when done.

**Comment by dalto8** [comment]
> You should rebase your `non-EOS` branch onto `work-3.3` and force-push it (the `non-EOS` branch) when done.

I rebased off the work-3.3 and force pushed the changes but github is still showing 21 changed files so I may have missed something.

Any tips on how to clean this up from a git perspective would be appreciated.

**Comment by kkofler** [comment]
You can do an interactive rebase, in which you can skip unwanted commits and retain only the relevant ones.

**Comment by dalto8** [comment]
> You can do an interactive rebase, in which you can skip unwanted commits and retain only the relevant ones.

Thanks for the advice.  That cleaned it up.

I think this is ready for review now.

File: src/modules/bootloader/bootloader.conf
Comment (adriaandegroot) [inline src/modules/bootloader/bootloader.conf]: If this is supposed to be a list? Why not make it a list in YAML then? Notation is with bullets or `[ ]`, and then you get a list in the job-configuration, rather than a string that you need to split and handle separately.

File: src/modules/bootloader/main.py
Comment (adriaandegroot) [inline src/modules/bootloader/main.py:86]: Thank you for useful docstrings.

**Review by adriaandegroot — COMMENTED** [review]
LGTM, with the one this-could-be-a-list as a non-blocking suggestion for improvement.

**Review by adriaandegroot — APPROVED** [review]

**Comment by dalto8** [comment]
I made that last change you requested and rebased it again.  Let me know if there is anything that needs to be done before it can be merged.

**Pull request merged**
